### PR TITLE
Output Tensor Shape Validation b/w ONNX inference and ORT 

### PR DESCRIPTION
**Pull request by sumitsays**
**Background**:
During the execution of a model on CPU, _output tensor shape generated by ONNX inference_ needs to be validated against the  _output tensor shape generated by ORT_. 
Today, this validation happens only if the space for output tensor is pre-allocated. But, to make the behavior consistent with GPU execution provider (DML), this validation needs to happen irrespective of the space pre-allocation.

**Motivation**:
Bug raised by a user, when a in-consistency has been seen b/w the CPU ep and GPU ep. [Github Bug: 6075](https://github.com/microsoft/onnxruntime/issues/6075) ,

 **Implementation**:
Before the actual execution of a operator on cpu, this validation will happen. As of now, if this validation fails, it will display/log a warning message stating the validation failure and continue with the execution. But ideally to make it consistent with GPU behavior, ORT should also stop the execution then and there if the validation fails. It is done like this for now because there are multiples existing tests fail because of this validation failure. First those tests need to be fixed, then ORT should throw an exception and stop the execution. Known tests which were failing because of validation failure:
•	PadOpTest.Pad_Edge_DimWithZeroInput 
•	ReductionOpTest.ReduceDimWithZero
•	Loop.SubgraphInputShadowsOuterScopeValue

There were more tests which were failing, but actual reason needs to be investigated. Few of them were:
•	CudaKernelTest.LayerNorm_SmallSizeTensor_IntermediateAxis 
•	CudaKernelTest.SimplifiedLayerNorm_SmallSizeTensor_IntermediateAxis
This was the [Azure build](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=349705&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=b24cf96b-e476-5cd6-97ae-776d6935d22f&l=8050) but it is not present anymore in the pipeline.

### CODE REVIEW

File: onnxruntime/core/optimizer/optimizer_execution_frame.cc [hunk 1]
```diff
@@ -86,6 +86,10 @@ OptimizerExecutionFrame::OptimizerExecutionFrame(const Info& info, const std::ve
   Init(std::vector<int>(), std::vector<OrtValue>(), info.GetInitializers(), std::vector<OrtValue>());
 }
 
+const logging::Logger* OptimizerExecutionFrame::GetLogger() {
+  return NULL;
```
Comment (fdwr) [inline onnxruntime/core/optimizer/optimizer_execution_frame.cc:90]: nullptr

File: onnxruntime/core/framework/execution_frame.cc [hunk 2]
```diff
@@ -86,13 +86,54 @@ Status IExecutionFrame::GetOrCreateNodeOutputMLValue(int index, const TensorShap
...
     } else {
+      VerifyOutputSizes(output_index, node, shape);
       status = CreateNodeOutputMLValueImpl(*p_ort_value, ort_value_idx, shape, nnz);
     }
   }
 
   return status;
 }
 
+void IExecutionFrame::VerifyOutputSizes(int output_index, const onnxruntime::Node& node,
```
Comment (pranavsharma) [inline onnxruntime/core/framework/execution_frame.cc:97]: onnxruntime:: is redundant. Several other places as well.

File: onnxruntime/core/framework/execution_frame.cc [hunk 3]
```diff
@@ -86,13 +86,54 @@ Status IExecutionFrame::GetOrCreateNodeOutputMLValue(int index, const TensorShap
...
+    return;
+  }
+
+  if (outputproto->value_case() != onnx::TypeProto::kTensorType) {
+    return;
+  }
+  const auto& tensortype = outputproto->tensor_type();
+  if (tensortype.has_shape()) {
+    const auto& shape = tensortype.shape();
+    if (shape.dim_size() != int(output_shape->NumDimensions())) {
```
Comment (pranavsharma) [inline onnxruntime/core/framework/execution_frame.cc:114]: use static_cast<int>

File: onnxruntime/core/framework/execution_frame.cc [hunk 4]
```diff
@@ -86,13 +86,54 @@ Status IExecutionFrame::GetOrCreateNodeOutputMLValue(int index, const TensorShap
...
+      if (GetLogger() != NULL) {
+        LOGS(*GetLogger(), WARNING) << "Number of dimension of output shape didn't match "
```
Comment (pranavsharma) [inline onnxruntime/core/framework/execution_frame.cc:116]: Log the actual and expected dims for easier debugging.

File: onnxruntime/core/framework/execution_frame.cc [hunk 5]
```diff
@@ -86,13 +86,54 @@ Status IExecutionFrame::GetOrCreateNodeOutputMLValue(int index, const TensorShap
...
+                                               << "with model's expected number of dimension of output shape";
+      }
+      return;
+    }
+
+    for (uint32_t output_dim = 0; output_dim < output_shape->NumDimensions(); ++output_dim) {
```
Comment (pranavsharma) [inline onnxruntime/core/framework/execution_frame.cc:122]:
```suggestion
    for (uint32_t output_dim = 0, end = output_shape->NumDimensions(); output_dim < end; ++output_dim) {
```

File: onnxruntime/core/framework/execution_frame.cc [hunk 6]
```diff
@@ -86,13 +86,54 @@ Status IExecutionFrame::GetOrCreateNodeOutputMLValue(int index, const TensorShap
...
+      if (shape.dim(output_dim).has_dim_value()) {
+        int64_t expected_size = shape.dim(output_dim).dim_value();
```
Comment (pranavsharma) [inline onnxruntime/core/framework/execution_frame.cc:124]: Can avoid calling shape.dim(output_dim) the second time.

File: onnxruntime/core/framework/execution_frame.cc [hunk 7]
```diff
@@ -86,13 +86,54 @@ Status IExecutionFrame::GetOrCreateNodeOutputMLValue(int index, const TensorShap
                     " Requested shape:", shape ? shape->ToString() : "null");
       }
     } else {
+      VerifyOutputSizes(output_index, node, shape);
```
Comment (pranavsharma) [inline onnxruntime/core/framework/execution_frame.cc:89]: I would rather do this check conditionally than paying the price *always*. An invalid ONNX model is probably not the norm and we shouldn't penalize the normal use case. Note that this verification is done for every single output index for every single node for every single kernel in the graph.

File: onnxruntime/core/optimizer/optimizer_execution_frame.cc (hunk 1 above)
Comment (sumitsays) [reply onnxruntime/core/optimizer/optimizer_execution_frame.cc:90]: Get it. Changed.

File: onnxruntime/core/framework/execution_frame.cc (hunk 2 above)
Comment (sumitsays) [reply onnxruntime/core/framework/execution_frame.cc:97]: Removed.

File: onnxruntime/core/framework/execution_frame.cc (hunk 3 above)
Comment (sumitsays) [reply onnxruntime/core/framework/execution_frame.cc:114]: Make sense. Done.

File: onnxruntime/core/framework/execution_frame.cc (hunk 4 above)
Comment (sumitsays) [reply onnxruntime/core/framework/execution_frame.cc:116]: Added the dims value.

File: onnxruntime/core/framework/execution_frame.cc (hunk 5 above)
Comment (sumitsays) [reply onnxruntime/core/framework/execution_frame.cc:122]: Got it. storing the number of dimensions in a variable will avoid the multiple call to the function.

File: onnxruntime/core/framework/execution_frame.cc (hunk 6 above)
Comment (sumitsays) [reply onnxruntime/core/framework/execution_frame.cc:124]: Acknowledged.

File: onnxruntime/core/framework/execution_frame.cc (hunk 7 above)
Comment (sumitsays) [reply onnxruntime/core/framework/execution_frame.cc]: Added a check for whether a node is an internal or final output node, before calling the verification method.

File: onnxruntime/core/framework/execution_frame.cc [hunk 8]
```diff
@@ -86,13 +86,61 @@ Status IExecutionFrame::GetOrCreateNodeOutputMLValue(int index, const TensorShap
...
+  }
+
+  if (outputproto->value_case() != onnx::TypeProto::kTensorType) {
+    return;
+  }
+  const auto& tensortype = outputproto->tensor_type();
+  if (tensortype.has_shape()) {
+    const auto& shape = tensortype.shape();
+
+    int actualNumDims = static_cast<int>(output_shape->NumDimensions());
```
Comment (hariharans29) [inline onnxruntime/core/framework/execution_frame.cc:117]: nit: camel casing style ?

Comment (sumitsays) [reply onnxruntime/core/framework/execution_frame.cc]: Changed.

File: onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc [hunk 9]
```diff
@@ -2277,12 +2283,12 @@ TEST(ReductionOpTest, ReduceDimWithZero) {
   };
 
   // reduce on all axes keeping dims. should allow the 0 to be the reduced value
-  OpTester test("ReduceSum", 10);
-  test.AddAttribute("keepdims", int64_t(1));
-  test.AddShapeToTensorData(true, 1);  // make second dim symbolic so that we don't break during shape inferencing
-  test.AddInput<float>("data", {3, 0, 2}, {});
-  test.AddOutput<float>("reduced", {1, 0, 1}, {});
-  run(test);
+  //OpTester test("ReduceSum", 10);
```
Comment (pranavsharma) [inline onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc:2286]: What's the plan for re-enabling these tests?

Comment (sumitsays) [reply onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc]: As discussed, enabled the test.

**Review by pranavsharma — DISMISSED** [review]

**Review by fdwr — DISMISSED** [review]
Reapproving. ORT already signed off, but Sumit encountered a bad merge and rewound it to an earlier commit.

**Review by fdwr — APPROVED** [review]
Reapproving after bad merge fixed (duplicate OptimizerExecutionFrame::GetLogger and test case name) and CI's pass. No other changes since ORT (Pranav and Hari) last signed off.

**Pull request merged**